pkg/controller/networkpolicy/mutate.go

```diff
@@ -0,0 +1,201 @@
+	"k8s.io/klog"
+
+	secv1alpha1 "github.com/vmware-tanzu/antrea/pkg/apis/security/v1alpha1"
+)
+
+type NetworkPolicyMutator struct {
+	networkPolicyController *NetworkPolicyController
+}
+
+// NewNetworkPolicyMutator returns a new *NetworkPolicyMutator.
+func NewNetworkPolicyMutator(networkPolicyController *NetworkPolicyController) *NetworkPolicyMutator {
+	return &NetworkPolicyMutator{
+		networkPolicyController: networkPolicyController,
+	}
+}
+
+// Mutate function mutate a Antrea Policy object
+func (m *NetworkPolicyMutator) Mutate(ar *admv1.AdmissionReview) *admv1.AdmissionResponse {
+	var result *metav1.Status
+	var msg string
+	var patch []byte
+	allowed := false
+	patchType := admv1.PatchTypeJSONPatch
+
+	op := ar.Request.Operation
+	curRaw := ar.Request.Object.Raw
+	oldRaw := ar.Request.OldObject.Raw
+
+	switch ar.Request.Kind.Kind {
+	case "ClusterNetworkPolicy":
+		klog.V(2).Info("Mutating Antrea ClusterNetworkPolicy CRD")
+		var curACNP, oldACNP secv1alpha1.ClusterNetworkPolicy
+		if curRaw != nil {
+			if err := json.Unmarshal(curRaw, &curACNP); err != nil {
+				klog.Errorf("Error de-serializing current Antrea ClusterNetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldACNP); err != nil {
+				klog.Errorf("Error de-serializing old Antrea ClusterNetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		msg, allowed, patch = m.mutateAntreaPolicyRuleName(op, curACNP.Spec.Ingress, curACNP.Spec.Egress)
+	case "NetworkPolicy":
+		klog.V(2).Info("Mutating Antrea NetworkPolicy CRD")
+		var curANP, oldANP secv1alpha1.NetworkPolicy
+		if curRaw != nil {
+			if err := json.Unmarshal(curRaw, &curANP); err != nil {
+				klog.Errorf("Error de-serializing current Antrea NetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldANP); err != nil {
+				klog.Errorf("Error de-serializing old Antrea NetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		msg, allowed, patch = m.mutateAntreaPolicyRuleName(op, curANP.Spec.Ingress, curANP.Spec.Egress)
+	}
+
+	if msg != "" {
+		result = &metav1.Status{
+			Message: msg,
+		}
+	}
+	return &admv1.AdmissionResponse{
+		Allowed:   allowed,
+		Result:    result,
+		PatchType: &patchType,
+		Patch:     patch,
+	}
+}
+
+// mutateAntreaPolicyRuleName mutates names of rules of an Antrea NetworkPolicy CRD.
+// If users didn't specify the name of an ingress or egress rule,
+// mutateAntreaPolicyRuleName will auto-generate a name for this rule.
+func (m *NetworkPolicyMutator) mutateAntreaPolicyRuleName(op admv1.Operation, ingress, egress []secv1alpha1.Rule) (string, bool, []byte) {
+	allowed := true
+	reason := ""
+	var patch []byte
+	switch op {
+	case admv1.Create, admv1.Update:
+		ruleNames := sets.NewString()
+		for _, rule := range append(ingress, egress...) {
+			ruleNames.Insert(rule.Name)
+		}
```

Comment:
will `append` create a new list here? I don't know if there will be actual overhead, but maybe we can just have 2 loops :)

Reply:
As discussed here https://github.com/vmware-tanzu/antrea/pull/1330#discussion_r510973495, it seems acceptable. What's your opinion about the tradeoff between overhead and code redundant?

Reply:
@tnqn : should not the overhead of append() depend on the number of rules? If we have 1K rules, the overhead can be high?

Reply:
I wouldn't worry too much about code duplication here since there is only one statement in the loop's body. I agree with @jianjuns that the overhead just gets more significant for a large number of rules

Reply:
I proposed to use append or a function to reduce 9 lines of code duplication. Since it's just one line loop, a function will be overkill, I agree it could just be two loops for this case.

Reply:
again.. I'm not really keen on mutation webhook doing some sort of validation and also adding overhead of checks and multiple loops here.. specifically for this use case.. imo this should be a simple mutator. but if others agree with this approach then we can move on

Reply:
Got you.